lib/cloudhsm/cloudhsm_key_generator.rb

```diff
@@ -0,0 +1,70 @@
+require 'greenletters'
+
+class CloudhsmKeyGenerator
+  KEY_MGMT_UTIL = '/opt/cloudhsm/bin/key_mgmt_util'.freeze
+  SSL_CONF = 'config/openssl.conf'.freeze
+
+  def initialize
+    @kmu = Greenletters::Process.new(KEY_MGMT_UTIL, transcript: $stdout)
+    @kmu.start!
+    @kmu.wait_for(:output, /Command:/)
+  end
+
+  def generate_saml_key
+    saml_label = create_key_and_crt_files
+    login_to_hsm
+    wrapping_key_handle = generate_symmetric_wrapping_key
+    import_private_key(saml_label, wrapping_key_handle)
+    exit_hsm
+    saml_label
+  end
+
+  private
+
+  def create_key_and_crt_files
+    timestamp = Time.zone.now.strftime('%Y%m%d%H%M%S')
+    saml_label = Shellwords.shellescape("saml_#{timestamp}")
+    system "openssl req -x509 -nodes -sha256 -days 365 -newkey rsa:2048 " \
+           "-keyout #{saml_label}.key " \
+           "-out #{saml_label}.crt -config #{SSL_CONF}"
+    saml_label
+  end
+
+  def login_to_hsm
+    username, password = username_password
+    @kmu << "loginHSM -u CU -s #{username} -p #{password}\n"
+    @kmu.wait_for(:output, /SUCCESS/)
+  end
+
+  def generate_symmetric_wrapping_key
+    @kmu << "genSymKey -t 31 -s 16 -sess -l wrapping_key_for_import\n"
+    wrapping_key_handle = wait_for_wrapping_key_handle
+    @kmu.wait_for(:output, /SUCCESS/)
+    wrapping_key_handle
+  end
+
+  def wait_for_wrapping_key_handle
+    wrapping_key_handle = nil
+    @kmu.wait_for(:output, /Key Handle: \d+/) do |_process, matching|
+      matching.matched =~ /Key Handle: (\d+)/
+      wrapping_key_handle = Regexp.last_match[1]
+    end
+    wrapping_key_handle
+  end
+
+  def import_private_key(saml_label, wrapping_key_handle)
+    @kmu << "importPrivateKey -f #{saml_label}.key -l #{saml_label} -w #{wrapping_key_handle}\n"
+    @kmu.wait_for(:output, /SUCCESS/)
+  end
+
+  def exit_hsm
+    @kmu << 'exit'
+  end
+
+  def username_password
+    cloudhsm_pin = Figaro.env.cloudhsm_pin
```

Comment:
Is this expected to be in `application.yml`? I'm not sure it's a good idea for the default IDP CloudHSM user to have permissions to manage keys.